src/megatron/energon/flavors/webdataset/sharder.py

```diff
@@ -181,21 +231,78 @@ def shard_workers(
         global_workers = num_workers * worker_config.world_size
 
-        samples_per_worker = total_samples / global_workers
+        min_samples_per_worker = int(total_samples / global_workers)
+        num_workers_with_more_samples = total_samples % global_workers
+
+        # We are going to compute the samples assigned to each worker on the current rank.
+        # This is done in multiple steps.
+        # Some of these steps could be collapsed into one, but we keep them separate for clarity:
+        # 1. Compute the number of samples per global worker (rotated by rotation_offset,
+        #    typically given by previous datasets).
+        # 2. Permute the nuber of samples per global worker by a generalized bit reversal sequence
+        # 3. Given the sample counts, compute the start and end indices for each global worker
+        # 4. Extract the local worker sample assignments for the current rank.
+        # 5. Split the shards based on the start and end indices.
+
+        # 1. Let's compute it globally for all workers first
+        num_samples_per_global_worker = []
+        for global_worker_idx in range(global_workers):
+            if (
+                global_worker_idx - rotation_offset + global_workers
+            ) % global_workers < num_workers_with_more_samples:
+                # This worker gets one more sample
+                num_samples_per_global_worker.append(min_samples_per_worker + 1)
+            else:
+                # This worker gets the minimum number of samples
+                num_samples_per_global_worker.append(min_samples_per_worker)
+
+        # 2. Permute the number of samples per global worker
+        worker_bitrev_seq = cls._generalized_bit_reversal(global_workers)
+
+        # The worker_bitrev_seq is the order in which any remainder samples shall
+        # be assigned to workers.
+        # That means, the x-axis (array index) is the remainder sample index
+        # and the y-axis (value) is the global worker index.
+        # So we map the y (value) to the old global worker index from the linear sequence.
+
+        new_num_samples_per_global_worker = [-1] * global_workers
+        for old_worker_idx, new_worker_idx in enumerate(worker_bitrev_seq):
+            new_num_samples_per_global_worker[new_worker_idx] = num_samples_per_global_worker[
+                old_worker_idx
+            ]
+
+        num_samples_per_global_worker = new_num_samples_per_global_worker
+
+        # 3. Compute the global worker sample start and end indices
+        global_worker_sample_start_ends = []
+        cur_offset = 0
+        for global_worker_idx in range(global_workers):
+            next_offset = cur_offset + num_samples_per_global_worker[global_worker_idx]
+            global_worker_sample_start_ends.append((cur_offset, next_offset))
+            cur_offset = next_offset
 
-        # Let's look at the workers of the current local rank.
```

Comment:
I believe, this can be reverted now, as well as the `_split_shards` function. We already have speed issues with this part for lots of shards, so maybe rather keep it optimized, and not remove optimization here?